### PR TITLE
update to PUPPI adding eta dependence, some bug fixes, and mechanism …

### PR DESCRIPTION
**Pull request by nhanvtran**
update to PUPPI adding eta dependence, some bug fixes, and mechanism for doing PUPPI diagnostics
Bug fixes related to miniAOD conventions and tunings for Phys14, which will have to be updated after a new tuning.

### CODE REVIEW

File: CommonTools/PileupAlgos/interface/PuppiAlgo.h [hunk 1]
```diff
@@ -16,6 +16,7 @@ class PuppiAlgo{
   void   computeMedRMS(const unsigned int &iAlgo,const double &iPVFrac);
   //Get the Weight
   double compute(std::vector<double> const &iVals,double iChi2) const;
+  std::vector<float> alphas(){ return fPups; }
```
Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiAlgo.h]:
```C++
const std::vector<float>& alphas() const { return fPups; }
```

File: CommonTools/PileupAlgos/interface/PuppiAlgo.h [hunk 2]
```diff
@@ -25,6 +26,9 @@ class PuppiAlgo{
   inline bool   isCharged  ( unsigned int iAlgo) const { return fCharged.at(iAlgo); }
   inline double coneSize  ( unsigned int iAlgo) const { return fConeSize.at(iAlgo); }
   inline double neutralPt  (int iNPV) const { return fNeutralPtMin + iNPV * fNeutralPtSlope; }
+  std::vector<double> fRMS;
```
Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiAlgo.h]: please move fRMS and fMedian to private: section

File: CommonTools/PileupAlgos/interface/PuppiContainer.h [hunk 3]
```diff
@@ -39,12 +39,20 @@ class PuppiContainer{
...
     std::vector<fastjet::PseudoJet> const & pvParticles() const { return fChargedPV; }        
-    std::vector<double> const & puppiWeights() ;
+    std::vector<double> const & puppiWeights();
+    std::vector<double> const & puppiRawAlphas(){ return fRawAlphas; }
+    std::vector<double> const & puppiAlphas(){ return fVals; }
+    // const std::vector<double> puppiAlpha   () {return fAlpha;}
+    std::vector<double> const & puppiAlphasMed() {return fAlphaMed;}
+    std::vector<double> const & puppiAlphasRMS() {return fAlphaRMS;}
+
+    int puppiNAlgos(){ return fNAlgos; }
```
Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiContainer.h]: all the new methods should be const-qualified (``` puppiNAlgos() const ``` )

Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiContainer.h]: I mean, the ones that don't change the state of PuppiContainer, should be const-qualified

File: CommonTools/PileupAlgos/src/PuppiAlgo.cc [hunk 4]
```diff
@@ -8,143 +8,221 @@
...
+    //if(!fCharged[iAlgo]) std::cout << " Process : " << iAlgo  << " Median : " << fMedian[iAlgo] << " +/- " << fRMS[iAlgo]  << " -- Begin : " << lNBefore << " -- Total :  " << fNCount[iAlgo] << " -- 50% " << lNHalfway  << " Fraction less than @ Median : " << std::endl;
+    
+    // if(!fAdjust[iAlgo]) return;
+    // //Adjust the p-value to correspond to the median
+    // std::sort(fPupsPV.begin(),fPupsPV.end());
+    // int lNPV = 0; for(unsigned int i0 = 0; i0 < fPupsPV.size(); i0++) if(fPupsPV[i0] <= lMed ) lNPV++;
+    // double lAdjust = 1.5*double(lNPV)/double(fPupsPV.size()+fNCount[iAlgo]);
+    // if(lAdjust > 0) fMedian[iAlgo] -= sqrt(ROOT::Math::chisquared_quantile(lAdjust,1.)*fRMS[iAlgo]);
 }
+////////////////////////////////////////////////////////////////////////////////
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiAlgo.cc]: unless all of this commented out code is going to be imminently necessary, please remove.
One can always go back to an older release or an older commit to make comparisons

File: CommonTools/PileupAlgos/src/PuppiContainer.cc [hunk 5]
```diff
@@ -12,194 +12,243 @@ using namespace std;
...
-  }
-  if (fPVFrac != 0) fPVFrac = double(fChargedPV.size())/fPVFrac;
-  else fPVFrac = 0;
+    fPVFrac = 0.;
+    fNPV    = 1.;
+    fRecoParticles = iRecoObjects;
+    for (unsigned int i = 0; i < fRecoParticles.size(); i++){
+        fastjet::PseudoJet curPseudoJet;
+        auto fRecoParticle = fRecoParticles[i];
+        float nom = sqrt((fRecoParticle.m)*(fRecoParticle.m) + (fRecoParticle.pt)*(fRecoParticle.pt)*(cosh(fRecoParticle.eta))*(cosh(fRecoParticle.eta))) + (fRecoParticle.pt)*sinh(fRecoParticle.eta);//hacked
```
Comment (slava77) [inline CommonTools/PileupAlgos/src/PuppiContainer.cc]: why not add a rapidity() method to the RecoObj ?

File: CommonTools/PileupAlgos/test/testPUMods.py [hunk 6]
```diff
@@ -5,18 +5,25 @@
 process.load('FWCore/MessageService/MessageLogger_cfi')
 process.load('Configuration/StandardSequences/FrontierConditions_GlobalTag_cff')
 process.MessageLogger.cerr.FwkReport.reportEvery = 10
-process.GlobalTag.globaltag = 'START53_V7G::All'
+#process.GlobalTag.globaltag = 'MCRUN2_75::All'
+#process.GlobalTag.globaltag = 'auto:run1_mc'
+from Configuration.AlCa.GlobalTag import GlobalTag
+process.GlobalTag = GlobalTag(process.GlobalTag, 'auto:run1_mc', '')
```
Comment (slava77) [inline CommonTools/PileupAlgos/test/testPUMods.py]: you have a run2 input file. Better keep the GT consistent as well

File: CommonTools/PileupAlgos/interface/PuppiAlgo.h (hunk 2 above)
Comment (nhanvtran) [inline CommonTools/PileupAlgos/interface/PuppiAlgo.h]: Actually I need this to be public, it's used elsewhere

**Comment by nhanvtran** [comment]
Thanks for the comments Slava -- I've implemented all the changes but one which I disagree with.

File: CommonTools/PileupAlgos/interface/PuppiAlgo.h [hunk 7]
```diff
@@ -16,6 +16,7 @@ class PuppiAlgo{
   void   computeMedRMS(const unsigned int &iAlgo,const double &iPVFrac);
   //Get the Weight
   double compute(std::vector<double> const &iVals,double iChi2) const;
+  const std::vector<float> alphas(){ return fPups; }
```
Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiAlgo.h]: return const reference, not a copy of the vector

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 8]
```diff
@@ -191,6 +200,35 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
+    for (unsigned int i = 0; i < lRawAlphas.size(); i++){
+      alphas->push_back( lRawAlphas[i] );
+    }
+    // THE alpha per particle
+    std::auto_ptr<std::vector<double> > theAlphas(new std::vector<double>);
+    std::auto_ptr<std::vector<double> > theAlphasMed(new std::vector<double>);
+    std::auto_ptr<std::vector<double> > theAlphasRms(new std::vector<double>);
+    for (unsigned int i = 0; i < lAlphas.size(); i++){ theAlphas->push_back( lAlphas[i] ); }
+    for (unsigned int i = 0; i < lAlphasMed.size(); i++){ theAlphasMed->push_back( lAlphasMed[i] ); }
+    for (unsigned int i = 0; i < lAlphasRms.size(); i++){ theAlphasRms->push_back( lAlphasRms[i] ); }
```
Comment (slava77) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: I don't see any manipulations with data members here.
For this case, why not just 
```C++
std::auto_ptr<std::vector<double> > theAlphas(new std::vector<double>(fPuppiContainer->puppiAlphas()));
```
and similarly for everything else

File: CommonTools/PileupAlgos/interface/PuppiAlgo.h [hunk 9]
```diff
@@ -26,6 +27,9 @@ class PuppiAlgo{
   inline double coneSize  ( unsigned int iAlgo) const { return fConeSize.at(iAlgo); }
   inline double neutralPt  (int iNPV) const { return fNeutralPtMin + iNPV * fNeutralPtSlope; }
 
+  inline double RMS( unsigned int i ) const {return fRMS[i];}
+  inline double Median( unsigned int i ) const {return fMedian[i];}
```
Comment (slava77) [inline CommonTools/PileupAlgos/interface/PuppiAlgo.h]: method names should start with lower case letter (CMS style)

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/9142	ed928db


- event content in miniAOD is up by 0.9% (70 events only, maybe less on larger events)
- most of the increase is from slimmedJetsPuppi going up by 15%

```
-------------------------------------------------------------
 or, B       new, B      delta, B   delta, %   deltaJ, %    branch 
-------------------------------------------------------------
    18424 ->       18633        209      1.1   0.31     patPackedCandidates_packedPFCandidates__PAT.
     2217 ->        2572        355     14.8   0.53     patJets_slimmedJetsPuppi__PAT.
-------------------------------------------------------------
    66479 ->       67044        565             0.9     ALL BRANCHES

```

CPU is up by 25%, mainly in the puppi producer (x6 slower)
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.437546     +25.47%        75.80 ms/ev ->       463.24 ms/ev puppi
```

all of this is coming from 
- ~5 times more spent in PuppiContainer::getRMSAvg
- new PuppiContainer::getRawAlphas (with about the same time as the getRMSAvg)
The underlying problem is the use of PuppiContainer::var_within_R, which goes through fastjet internals.

@gpetruc do you approve 25% increase in miniAOD CPU?
I suggest some refactoring is done in time for pre6 (sooner is better).

**Pull request merged**